File: Documentation/4.1/Samples/csharp/Raven.Documentation.Samples/ClientApi/Session/Debugging/IncludeExplanations.cs
// -----------------------------------------------------------------------
//  <copyright file="IncludeExplanations.cs" company="Hibernating Rhinos LTD">
//      Copyright (c) Hibernating Rhinos LTD. All rights reserved.
//  </copyright>
// -----------------------------------------------------------------------

using System.Linq;
using System.Threading.Tasks;
using Raven.Client.Documents;
using Raven.Client.Documents.Queries.Explanation;
using Raven.Client.Documents.Session;
using Raven.Documentation.Samples.Orders;

namespace Raven.Documentation.Samples.ClientApi.Session.Debugging
{
    public class IncludeExplanations
    {
        public interface IFoo<T>
        {
            #region explain_1
            IDocumentQuery<T> IncludeExplanations(out Explanations explanations);
            #endregion
        }

        public async Task Explain()
        {
            using (var store = new DocumentStore())
            {
                using (var session = store.OpenSession())
                {
                    #region explain_2
                    var syrups = session.Advanced.DocumentQuery<Product>()
                        .IncludeExplanations(out Explanations explanations)
                        .Search(x => x.Name, "Syrup")
                        .ToList();

                    string[] scoreDetails = explanations.GetExplanations(syrups[0].Id);
                    #endregion
                }

                using (var session = store.OpenAsyncSession())
                {
                    #region explain_3
                    var syrups = await session.Advanced.AsyncDocumentQuery<Product>()
                        .IncludeExplanations(out Explanations explanations)
                        .Search(x => x.Name, "Syrup")
                        .ToListAsync();

                    string[] scoreDetails = explanations.GetExplanations(syrups[0].Id);
                    #endregion
                }
            }
        }
    }
}

File: Documentation/5.2/Samples/java/src/test/java/net/ravendb/ClientApi/Session/Querying/Debugging/IncludeExplanations.java
package net.ravendb.ClientApi.Session.Debugging;

import net.ravendb.client.documents.DocumentStore;
import net.ravendb.client.documents.IDocumentStore;
import net.ravendb.client.documents.queries.explanation.Explanations;
import net.ravendb.client.documents.session.IDocumentQuery;
import net.ravendb.client.documents.session.IDocumentSession;
import net.ravendb.client.primitives.Reference;
import net.ravendb.northwind.Product;

import java.util.List;

public class IncludeExplanations {

    public interface IFoo<T> {
        //region explain_1
        IDocumentQuery<T> includeExplanations(Reference<Explanations> explanations);
        //endregion
    }

    public void explain() {
        try (IDocumentStore store = new DocumentStore()) {
            try (IDocumentSession session = store.openSession()) {
                //region explain_2
                Reference<Explanations> explanationRef = new Reference<>();
                List<Product> syrups = session.advanced().documentQuery(Product.class)
                    .includeExplanations(explanationRef)
                    .search("Name", "Syrup")
                    .toList();

                String[] scoreDetails = explanationRef.value.getExplanations(syrups.get(0).getId());
                //endregion
            }
        }
    }
}
